Removes bootstrap and add support for Baseweb ui

balwant-jtc opened the pull request:
## Description
Removes bootstrap and add support for Baseweb ui



https://github.com/jalantechnologies/boilerplate-mern/assets/78337443/15897f82-6e1b-45dc-9a14-809a2d531d8d

jjalan commented on src/apps/frontend/app.component.tsx:
```diff
@@ -20,20 +19,22 @@ export default function App(): React.ReactElement {
-      </Router>
-    </DepsProvider>
+    <BaseWebProvider>
+      <DepsProvider
```
We should deprecate `DepsProvider`. We initially thought to use dependency injection and `DepsProvider` knows all about application dependencies, kinda inspired by angular. React has its own context paradigm which is better suited, and hence this should be deprecated

jjalan commented on src/apps/frontend/app.global.css:
```diff
@@ -0,0 +1,29 @@
+/* add your global imports and css rules her */
```
Where is this coming from?

I believe with baseweb, there should not be any need to write css like this, except when we work with 3rd part lib and we need to override some of the css directly (which again should be discouraged).

jjalan commented on src/apps/frontend/hooks/login-form.hook.ts:
```diff
@@ -0,0 +1,50 @@
+      setError(false);
+
+      try {
+        const accessService = new AccessService();
```
You can move the initialization of this outside this UI component i.e. 

```
const accessService = new AccessService();
const useLoginForm = () => {
```

Should we also move this closer to the login form, and avoid creating another folder `hooks`? I can see in future everyone would put all context here which would make this folder explode

also, `AccessService` seems mislead, recommend to rename to better reflect what it is.

Balwant-Shekhawat replied:
Done, renamed it to AuthService

jjalan commented on src/apps/frontend/hooks/login-form.hook.ts:
```diff
@@ -0,0 +1,50 @@
+      try {
+        const accessService = new AccessService();
+        await accessService.login(username, password);
+        setSuccess(true);
```
You want to consider storing the token and response into local storage, so that once user login, it can be read from it. You want to expose logout button to then clear local storage.

This can also be considered as a separate PR in case there is no logged in page today that exists after user login.

Balwant-Shekhawat replied:
Will do this after implementing [this](https://teams.microsoft.com/l/message/19:c388dc298f0f432bb4805cadc4ffe92f@thread.tacv2/1705932812553?tenantId=79836b2a-53cc-4854-81b4-ba2d7c9f2726&groupId=c0e79bd0-36ec-4b10-8be7-55f5b3d2395c&parentMessageId=1705932812553&teamName=Jalan%20Technologies&channelName=Engineering&createdTime=1705932812553)

jjalan commented on src/apps/frontend/pages/authentication/login/login-form.tsx:
```diff
@@ -0,0 +1,115 @@
+          </Checkbox>
+        </Block>
+        <Block>
+          <Button size={SIZE.default} overrides={{
+            BaseButton: {
+              style: () => ({
+                width: '100%',
+              }),
+            },
+          }}
```
hmm.

I assume there is no property on basweb that expands button to take full width? Does setting display type as block on the button also does not achieve that?

Balwant-Shekhawat replied:
No, it didn't worked

Balwant-Shekhawat replied:
But now handled it with the boolean prop `fullWidth`

jjalan commented on src/apps/frontend/pages/authentication/login/login-form.tsx:
```diff
@@ -0,0 +1,115 @@
+            onChange={formik.handleChange}
+            labelPlacement={LABEL_PLACEMENT.right}
+            overrides={{
+              Checkmark: {
+                style: ({ $theme }) => ({
+                  backgroundColor: `${$theme.colors.inputBackground}`,
+                  border: `1px solid ${$theme.border.borderColor}`,
+                  borderRadius: '0.25em',
+                }),
+              },
```
Override globally or create custom component with the overrides. A page should not create overrides for primitive components

Balwant-Shekhawat replied:
Created Custom Component

jjalan commented on src/apps/frontend/pages/authentication/login/login-form.tsx:
```diff
@@ -0,0 +1,115 @@
+            />
+          </FormControl>
+        </Block>
+        <Block style={{ marginTop: '1rem', marginBottom: '1rem' }}>
```
Use theme spacing class to control margins and paddings.

Also I believe you want to set rowGap on the Block that contains the input form to create consistent spacing between each bloc element.

Balwant-Shekhawat replied:
tried using `GridRowGap` but it didn't worked, so used theme spacing to make it work.

jjalan commented on src/apps/frontend/pages/authentication/login/login-form.tsx:
```diff
@@ -0,0 +1,115 @@
+            <Input
+              overrides={{
+                Root: {
+                  props: {
```
Remove please

Balwant-Shekhawat replied:
Done

jjalan commented on src/apps/frontend/pages/authentication/login/login-form.tsx:
```diff
@@ -0,0 +1,115 @@
+                    'data-testid': 'password',
+                  },
+                },
+                Input: {
```
Override globally or create custom component with the overrides. A page should not create overrides for primitive components

Balwant-Shekhawat replied:
Created custom component

jjalan commented on src/apps/frontend/pages/authentication/login/login-form.tsx:
```diff
@@ -0,0 +1,115 @@
+              overrides={{
+                Root: {
+                  props: {
+                    'data-testid': 'username',
```
Remove

Balwant-Shekhawat replied:
Done

jjalan commented on src/apps/frontend/pages/authentication/login/login-form.tsx:
```diff
@@ -0,0 +1,115 @@
+                  },
+                },
+                Input: {
+                  style: ({ $theme }) => ({
```
Override globally or create custom component with the overrides. A page should not create overrides for primitive components

Balwant-Shekhawat replied:
Created Custom Component

jjalan commented on src/apps/frontend/hooks/login-form.hook.ts:
```diff
@@ -0,0 +1,50 @@
+import { AccessService } from '../services';
+
+const useLoginForm = () => {
+  const [success, setSuccess] = useState(false);
+  const [error, setError] = useState(false);
```
Recommend creating [this](https://teams.microsoft.com/l/message/19:c388dc298f0f432bb4805cadc4ffe92f@thread.tacv2/1705932812553?tenantId=79836b2a-53cc-4854-81b4-ba2d7c9f2726&groupId=c0e79bd0-36ec-4b10-8be7-55f5b3d2395c&parentMessageId=1705932812553&teamName=Jalan%20Technologies&channelName=Engineering&createdTime=1705932812553) abstraction

Balwant-Shekhawat replied:
Will try to come up with it, but for now resuming on Valon's tasks.

jjalan commented on src/apps/frontend/hooks/index.ts:
```diff
@@ -0,0 +1,4 @@
+/* eslint-disable import/prefer-default-export */
```
Can we either consider updating lint rules or follow the guidelines so we don't have to suppress here and other places?

Balwant-Shekhawat replied:
I choose to suppress the warning here because in the future, we might want to export multiple named items from the index.ts file. Currently, we're only exporting a single hook, and TypeScript is reminding us to use the default export. However, if we comply with this warning now, it could lead to unnecessary changes throughout the codebase when we decide to export multiple items later on.

And if we chose to suppress the warning from the TypeScript configuration. This way, we won't be alerted even if we later export just a single item from a file. Which is also not correct.

While generally not a fan of suppressing warnings, in this specific situation, it seems reasonable to do so for the sake of future flexibility and to prevent unnecessary code modifications.

 Interestingly, Baseweb's codebase also includes instances where these TypeScript warnings are suppressed.

jjalan commented on src/apps/frontend/pages/authentication/login/login-form.tsx:
```diff
@@ -0,0 +1,137 @@
+      <Block>
+        <Button
+          size={SIZE.default}
+          overrides={{
+            BaseButton: {
+              style: () => ({
+                width: '100%',
+              }),
+            },
+          }}
```
For now, let's remove this and center the button.

Balwant-Shekhawat replied:
Used a boolean prop to handle this

jjalan commented on src/apps/frontend/pages/authentication/login/login-form.tsx:
```diff
@@ -0,0 +1,137 @@
+          checked={formik.values.rememberMe}
+          onChange={formik.handleChange}
+          labelPlacement={LABEL_PLACEMENT.right}
+          overrides={{
+            Checkmark: {
+              style: ({ $theme }) => ({
+                backgroundColor: `${$theme.colors.inputBackground}`,
+                border: `1px solid ${$theme.border.borderColor}`,
+                borderRadius: '0.25em',
+              }),
+            },
+            Label: {
+              style: ({ $theme }) => ({
+                color: `${$theme.colors.fontColor}`,
+              }),
+            },
+          }}
```
No instance based overrides please

Balwant-Shekhawat replied:
Added just to align with the style of the page, if you want i can remove the override part but then it will look different from the page.

jjalan commented on src/apps/frontend/hooks/login-form.hook.ts:
```diff
@@ -0,0 +1,50 @@
+        .email(constant.EMAIL_VALIDATION_ERROR)
+        .required('Required'),
+      password: Yup.string()
+        .min(constant.PASSWORD_MIN_LENGTH, constant.PASSWORD_VALIDATION_ERROR)
```
Please use this for password validation - https://github.com/dropbox/zxcvbn

jjalan commented on src/apps/frontend/types/async-operation.ts:
```diff
@@ -0,0 +1,18 @@
+
+export interface AsyncOperation<T> {
+  result?: T;
+  error?: ServiceError;
+  status: AsyncOperationStatus;
```
Async operation is an upper level abstraction than Service Error. This does not sound right

balwant-jtc replied:
Done

jjalan commented on src/apps/frontend/types/async-operation.ts:
```diff
@@ -0,0 +1,18 @@
+import { ServiceError, ServiceResponse } from './service-response';
+
+export enum AsyncOperationStatus {
+  Idle = 'idle',
+  Pending = 'pending',
+  Success = 'success',
+  Error = 'error',
```
Should they be all uppercase?

balwant-jtc replied:
Done

jjalan commented on src/apps/frontend/pages/authentication/login/login-form.tsx:
```diff
@@ -0,0 +1,75 @@
+
+  return (
+    <form onSubmit={formik.handleSubmit}>
+      {status !== AsyncOperationStatus.Idle ? (
```
This does not sound right - I think if operation is success, then it should show success

balwant-jtc replied:
Done

jjalan reviewed (commented):
Can you add a demo video of the new login experience?

jjalan commented on src/apps/frontend/types/async-operation.ts:
```diff
@@ -0,0 +1,18 @@
+  Error = 'error',
+}
+
+export interface AsyncOperation<T> {
```
`K, AsyncError<Y>`

balwant-jtc replied:
Addressed

balwant-jtc commented on src/apps/frontend/services/auth.service.ts:
```diff
@@ -0,0 +1,21 @@
+  login = async (
+    username: string,
+    password: string,
+  ): Promise<ServiceResponse<AccessToken>> => {
```
You mean `Promise<AsyncResponse<AccessToken, ApiError>>`? 
I can do this by doing `class ServiceResponse implements AsyncResponse` but then `AsyncResponse` has a `status` property. The service layer won't handle the status; instead, it will be managed by the `useAsync` hook, which sets it to values like `IDLE, PENDING, SUCCESS, or ERROR`."

jjalan replied:
> The service layer won't handle the status

Should it?

balwant-jtc replied:
IMO No, because even if service change the status from Pending to Sucess or Error, UI won't get updated. Also then `useAsync` would be of no use as it cant be used inside a service.

balwant-jtc replied:
The service layer should not have dependencies or concerns related to the UI layer.

jjalan replied:
I think `status` is not a concern of UI layer, rather its an inherent property of any async operation.

balwant-jtc replied:
But then it should be limited to SUCCESS or ERROR only. And there will be a middle layer(hook) which will expose `loading: boolean` to UI layer

balwant-jtc commented on src/apps/frontend/components/checkbox/checkbox.component.tsx:
```diff
@@ -0,0 +1,39 @@
+import React from 'react';
+
+export interface CheckboxProps extends BaseCheckboxProps {
+  testId: string;
```
Done

balwant-jtc commented on src/apps/frontend/components/button/button.component.tsx:
```diff
@@ -0,0 +1,29 @@
+import React from 'react';
+
+export interface ButtonProps extends BaseButtonProps {
+  testId?: string;
```
Done

jjalan commented on src/apps/frontend/app.global.css:
```diff
@@ -0,0 +1,14 @@
+* {
```
We want to avoid this file

balwant-jtc replied:
The purpose of including this file is to manage the margin and padding of the body, a task that, in my opinion, BaseWeb cannot handle.

jjalan replied:
Possibly yes, but this opens up and can often act as a reference that its ok to include raw css

Balwant-Shekhawat replied:
Added a note saying `AVOID ADDING MORE CSS HERE. UTILIZE BASEWEB STYLING INSTEAD` in this file but didn't removed this file because i didn't find a way to change margin padding of body through baseweb

jjalan replied:
What happens if we don't remove it for time being?

jjalan commented on src/apps/frontend/components/form-input/form-input.component.tsx:
```diff
@@ -0,0 +1,23 @@
+    overrides={{
+      Label: {
+        style: ({ $theme }) => ({
+          color: `${$theme.colors.fontColor}`,
```
Why do we need to override this? Can we not and use the form control directly at consumer?

Balwant-Shekhawat replied:
Done

jjalan commented on src/apps/frontend/components/input/input.component.tsx:
```diff
@@ -0,0 +1,34 @@
+        props: {
+          'data-testid': testId,
+        },
+        style: ({ $theme }) => ({
+          backgroundColor: `${$theme.colors.inputBackground}`,
+          color: `${$theme.colors.fontColor}`,
+          caretColor: `${$theme.colors.fontColor}`,
+        }),
```
Why do we need to override this? Can we not and use the form control directly at consumer?

Balwant-Shekhawat replied:
Done

jjalan commented on src/apps/frontend/contexts/async.hook.ts:
```diff
@@ -0,0 +1,45 @@
+} from '../types';
+
+const useAsync = <T>(): UseAsyncReturn<T> => {
+  const [asyncOperation, setAsyncOperation] = useState<
```
rename to `asyncOperationResult`

Balwant-Shekhawat replied:
Refactored the hook

jjalan commented on src/apps/frontend/pages/authentication/login/login-form.hook.ts:
```diff
@@ -0,0 +1,49 @@
+
+const useLoginForm = () => {
+  const { login: signIn } = useAuthContext();
+  const {
+    status, error, result, apiCallback,
+  } = useAsync();
+
```
This semantic seems a bit odd.

I would expect it to be:

```
const {
    status, error, result, asyncOperation: as login
  } = useAsync(signIn(username, password)));
```

This should not be needed:

```
const login = useCallback(
    ({ username, password }) => {
      apiCallback(() => signIn(username, password));
    },
    [apiCallback, signIn],
  );
```

Balwant-Shekhawat replied:
refactored it.

jjalan commented on src/apps/frontend/theme.ts:
```diff
@@ -0,0 +1,23 @@
+export const primitives = {
+  accent: '#0d6efd',
+  accent600: '#0b5ed7',
+  mono100: '#495057',
+  white: '#dee2e6',
+  mono300: '#2B3035',
+  mono400: '#212529',
+  negative: 'red',
+};
+
+export const overrides = {
+  colors: {
+    buttonPrimaryFill: primitives.accent,
+    buttonPrimaryHover: primitives.accent600,
+    backgroundPrimary: primitives.mono300,
+    fontColor: primitives.white,
+    inputBackground: primitives.mono400,
+    inputPlaceholder: primitives.white,
+  },
+  border: {
+    borderColor: primitives.mono100,
+  },
+};
```
For now, can we not create any overrides, and use whatever comes from basweb?

balwant-jtc replied:
Removed the primitives and overrides, and used default styling.

jjalan commented on src/apps/frontend/types/async-operation.ts:
```diff
@@ -0,0 +1,18 @@
+}
+
+export interface AsyncResult<T, AsyncError> {
+  result?: T;
+  error?: AsyncError;
+  status: AsyncOperationStatus;
+}
```
order alphabetically

Balwant-Shekhawat replied:
Done

jjalan commented:
1. Can we keep login page background white?
2. As you soon as you enter username, password field also showed error, this is not desireable as user did not even worked with that field.
3. The word `Success` is ugly. Can we use a toast? This is - I am assuming there is no logged in page right now which shows a header with username and ability to logout.

balwant-jtc commented:
> 1. Can we keep login page background white?
> 2. As you soon as you enter username, password field also showed error, this is not desireable as user did not even worked with that field.
> 3. The word `Success` is ugly. Can we use a toast? This is - I am assuming there is no logged in page right now which shows a header with username and ability to logout.

Addressed all points.

balwant-jtc commented:
> 1. Can we keep login page background white?
> 2. As you soon as you enter username, password field also showed error, this is not desireable as user did not even worked with that field.
> 3. The word `Success` is ugly. Can we use a toast? This is - I am assuming there is no logged in page right now which shows a header with username and ability to logout.

Addressed all points.

bbbneo333 commented on src/apps/frontend/pages/authentication/login/login-form.tsx:
```diff
@@ -0,0 +1,84 @@
+/* eslint-disable @typescript-eslint/restrict-template-expressions */
```
We are disabling this in multiple files. What is about and can we look into addressing it?

jjalan commented on src/apps/frontend/components/button/button.component.tsx:
```diff
@@ -0,0 +1,27 @@
+/* eslint-disable @typescript-eslint/restrict-template-expressions */
```
Can we please avoid this?

balwant-jtc replied:
Done

jjalan commented on src/apps/frontend/pages/authentication/login/login-form.hook.ts:
```diff
@@ -0,0 +1,43 @@
+import useAsync from '../../../contexts/async.hook';
+
+const useLoginForm = () => {
+  const { login: signIn } = useAuthContext();
+  const {
+    isLoading,
+    error,
+    result,
+    apiCallback: loginCallback,
+  } = useAsync(signIn);
```
`const { isLoginLoading, loginError, loginResult, login} = useAuthContext();`

balwant-jtc replied:
Done

jjalan commented on src/apps/frontend/pages/authentication/login/login-form.tsx:
```diff
@@ -0,0 +1,84 @@
+/* eslint-disable @typescript-eslint/restrict-template-expressions */
```
Please find a way to remove this

balwant-jtc replied:
Done

jjalan commented on src/apps/frontend/pages/authentication/login/login-form.tsx:
```diff
@@ -0,0 +1,84 @@
+
+import useLoginForm from './login-form.hook';
+
+export default function LoginForm(): React.ReactElement {
```
`export default LoginForm(): React.ReactElement = () => {`

balwant-jtc replied:
Done

jjalan commented on src/apps/frontend/pages/authentication/login/login-form.tsx:
```diff
@@ -0,0 +1,84 @@
+
+export default function LoginForm(): React.ReactElement {
+  const {
+    formik, isLoading, error, result,
```
`form`

Later, it should probably be used as:

`<form {...form}>`

balwant-jtc replied:
Done

jjalan commented on src/apps/frontend/pages/authentication/login/index.tsx:
```diff
@@ -0,0 +1,16 @@
+export const Login: React.FC = () => (
+  <AuthenticationPageLayout>
+    <AuthenticationFormLayout>
+      <LoginForm></LoginForm>
```
`onSuccess`, `onError`.

The page should be responsible to show error if its an error via toast, or navigate to `/` (or show toast for now) if success

jjalan replied:
`<LoginForm onSuccess ={onSuccess} onError ={onError}></LoginForm>`

jjalan commented on src/apps/frontend/types/async-operation.ts:
```diff
@@ -0,0 +1,11 @@
+  result?: T;
+}
+
+export interface UseAsyncResponse<T> extends AsyncResult<T, ApiError> {
```
This is a high level construct, should not anything about api.

balwant-jtc replied:
Done

jjalan commented on src/apps/frontend/pages/authentication/login/login-form.tsx:
```diff
@@ -0,0 +1,84 @@
+    formik, isLoading, error, result,
+  } = useLoginForm();
+
+  useEffect(() => {
+    if (!isLoading && error) {
+      toaster.negative(error.message);
+    } else if (!isLoading && result) {
+      toaster.positive('Login Successful');
+    }
+  }, [isLoading, error, result]);
```
We should remove this.

balwant-jtc replied:
Done

jjalan commented on src/apps/frontend/pages/authentication/login/login-form.tsx:
```diff
@@ -0,0 +1,84 @@
+
+      <Block>
+        <FormControl
+          label={() => 'Email address'}
```
label="Email Address"

balwant-jtc replied:
Done

jjalan commented on src/apps/frontend/pages/authentication/login/login-form.tsx:
```diff
@@ -0,0 +1,84 @@
+          error={formik.touched.username && formik.errors.username}
+        >
+          <Input
+            overrides={{
+              Input: {
+                props: {
+                  'data-testid': 'username',
+                },
+              },
+            }}
+            name="username"
+            placeholder="Email address"
+            value={formik.values.username}
+            onChange={formik.handleChange}
+            onBlur={formik.handleBlur}
+            autoComplete="off"
```
order alphabetically.

balwant-jtc replied:
Done

balwant-jtc commented on src/apps/frontend/pages/authentication/login/login-form.tsx:
```diff
@@ -0,0 +1,84 @@
+        </FormControl>
+      </Block>
+      <Block>
+        <Button type="submit" fullWidth disabled={isLoading}>
```
Done, `type` is not a property in baseweb so removed it from form as well.

jjalan reviewed (commented):
Can you please update the video to show the error scenario as well, ex - say your laptop does not have internet connectivity or say the backend api throws an error?

jjalan commented on src/apps/frontend/contexts/async.hook.ts:
```diff
@@ -0,0 +1,55 @@
+    [asyncFn],
+  );
+
+  useEffect(() => {
+    if (immediate) {
+      // eslint-disable-next-line @typescript-eslint/no-floating-promises
+      asyncCallback();
+    }
+  }, [asyncCallback, immediate]);
```
Why is this needed? I see `immediate` as always false

balwant-jtc replied:
If we want the function to run when the component mounts, we can set this flag to true. That way, we don't have to call the asyncCallback ourselves.

balwant-jtc replied:
For instance, if we need to retrieve the profile information of the logged-in user, we can directly use:

`const {isLoading, error, result } = useAuthContext(userProfile)`

There's no requirement to write it like this:

```
const {isLoading, error, result, userProfileCallback } = useAuthContext(userProfile)

userProfileCallback()
```

balwant-jtc replied:
Not using it at the moment.

balwant-jtc commented on src/apps/frontend/pages/authentication/login/index.tsx:
```diff
@@ -0,0 +1,31 @@
+  return (
+    <AuthenticationPageLayout>
+      <AuthenticationFormLayout>
+        <LoginForm isLoginLoading={isLoginLoading} {...formik}></LoginForm>
```
then how can i achieve this? : https://github.com/jalantechnologies/boilerplate-mern/pull/92#discussion_r1477759442

jjalan commented on src/apps/frontend/types/async-operation.ts:
```diff
@@ -0,0 +1,39 @@
+
+export class AsyncError {
+  code: string;
+  httpStatusCode: number;
```
This is not right. AsyncError cannot have httpStatusCode

balwant-jtc replied:
Removed

jjalan commented on src/apps/frontend/contexts/async.hook.ts:
```diff
@@ -0,0 +1,50 @@
+
+        return response?.data;
+      } catch (e) {
+        const err = e?.response?.data?.httpStatusCode
+          ? new AsyncError(e?.response?.data)
+          : new DefaultAsyncError(e);
```
async hook should not know anything about httpStatusCode

balwant-jtc replied:
Removed

jjalan commented on src/apps/frontend/pages/authentication/login/login-form.tsx:
```diff
@@ -0,0 +1,85 @@
+
+interface LoginFormProps {
+  onSuccess: () => void;
+  onError: (error: string) => void;
```
Can we pass error object instead of string?

balwant-jtc replied:
Done

jjalan commented on src/apps/frontend/pages/authentication/login/login-form.tsx:
```diff
@@ -0,0 +1,85 @@
+  onError: (error: string) => void;
+}
+
+const LoginForm: React.FC<LoginFormProps> = ({ onError, onSuccess }) => {
```
I am not sure, can you confirm - 

When user submitted the form - the form controls are disabled, including login button as button also shows loading icon?

balwant-jtc replied:
Form controls are not disabled only login button is disabled with loading icon.

jjalan replied:
We should disable form controls as well. When the login is in process we do not want user to edit their information.

balwant-jtc replied:
Done
